lib/Dialect/FIRRTL/Transforms/InferDomains.cpp

```diff
@@ -0,0 +1,1521 @@
+            Term *term) {
+  term = find(term);
+  if (auto *var = dyn_cast<VariableTerm>(term)) {
+    out << "?" << idTable.get(var);
+    return;
+  }
+  if (auto *val = dyn_cast<ValueTerm>(term)) {
+    auto value = val->value;
+    auto [name, rooted] = getFieldName(FieldRef(value, 0), false);
+    out << name;
+    return;
+  }
+  if (auto *row = dyn_cast<RowTerm>(term)) {
+    bool first = true;
+    out << "[";
+    for (size_t i = 0, e = info.getNumDomains(); i < e; ++i) {
+      auto domainOp = info.getDomain(DomainTypeID{i});
+      if (!first) {
+        out << ", ";
+        first = false;
+      }
+      out << domainOp.getName() << ": ";
+      render(info, out, idTable, row->elements[i]);
+    }
+    out << "]";
+    return;
+  }
+}
+
+LogicalResult unify(Term *lhs, Term *rhs);
+
+LogicalResult unify(VariableTerm *x, Term *y) {
+  assert(!x->leader);
+  x->leader = y;
+  return success();
+}
+
+LogicalResult unify(ValueTerm *xv, Term *y) {
+  if (auto *yv = dyn_cast<VariableTerm>(y)) {
+    yv->leader = xv;
+    return success();
+  }
+  if (auto *yv = dyn_cast<ValueTerm>(y)) {
+    return success(xv == yv);
+  }
+  return failure();
+}
+
+// NOLINTNEXTLINE(misc-no-recursion)
+LogicalResult unify(RowTerm *lhsRow, Term *rhs) {
+  if (auto *rhsVar = dyn_cast<VariableTerm>(rhs)) {
+    rhsVar->leader = lhsRow;
+    return success();
+  }
+  if (auto *rhsRow = dyn_cast<RowTerm>(rhs)) {
+    assert(lhsRow->elements.size() == rhsRow->elements.size());
+    for (auto [x, y] : llvm::zip(lhsRow->elements, rhsRow->elements)) {
+      if (failed(unify(x, y)))
+        return failure();
+    }
+    return success();
+  }
+
+  return failure();
+}
+
+// NOLINTNEXTLINE(misc-no-recursion)
+LogicalResult unify(Term *lhs, Term *rhs) {
+  if (!lhs || !rhs)
+    return success();
+  lhs = find(lhs);
+  rhs = find(rhs);
+  if (lhs == rhs)
+    return success();
+  if (auto *lhsVar = dyn_cast<VariableTerm>(lhs))
+    return unify(lhsVar, rhs);
+  if (auto *lhsVal = dyn_cast<ValueTerm>(lhs))
+    return unify(lhsVal, rhs);
+  if (auto *lhsRow = dyn_cast<RowTerm>(lhs))
+    return unify(lhsRow, rhs);
+  return failure();
+}
+
+void solve(Term *lhs, Term *rhs) {
+  [[maybe_unused]] auto result = unify(lhs, rhs);
+  assert(result.succeeded());
+}
+
+class TermAllocator {
+public:
+  /// Allocate a row of fresh domain variables.
+  RowTerm *allocRow(size_t size) {
+    SmallVector<Term *> elements;
+    elements.resize(size);
+    return allocRow(elements);
+  }
+
+  /// Allocate a row of terms.
+  RowTerm *allocRow(ArrayRef<Term *> elements) {
+    auto ds = allocArray(elements);
+    return alloc<RowTerm>(ds);
+  }
+
+  /// Allocate a fresh variable.
+  VariableTerm *allocVar() { return alloc<VariableTerm>(); }
+
+  /// Allocate a concrete domain.
+  ValueTerm *allocVal(Value value) { return alloc<ValueTerm>(value); }
+
+private:
+  template <typename T, typename... Args>
+  T *alloc(Args &&...args) {
```

Comment:
```suggestion
  [[nodiscard]] T *alloc(Args &&...args) {
```
?
Allocators are usually prime example of where this is useful -- that said we're using a bump pointer allocator so shrug.